service/history/tasks/state_machine_task.go

```diff
@@ -62,30 +63,32 @@ func (t *StateMachineTask) SetVisibilityTime(timestamp time.Time) {
 	t.VisibilityTimestamp = timestamp
 }
 
-// StateMachineCallbackTask is an outbound task on the callback queue.
-type StateMachineCallbackTask struct {
+// StateMachineOutboundTask is a task on the outbound queue.
+type StateMachineOutboundTask struct {
 	StateMachineTask
 	Destination string
 }
 
-func (t *StateMachineCallbackTask) GetDestination() string {
-	return t.Destination
+// GetDestination is used for grouping outbound tasks into a per source namespace and destination scheduler and in multi-cursor predicates.
+// Prefix the destination from the task with the task type to avoid collisions.
+func (t *StateMachineOutboundTask) GetDestination() string {
+	return fmt.Sprintf("type:%d:%s", t.StateMachineTask.Info.Type, t.Destination)
```

Comment:
I was debating between adding another attribute to the grouper and packing the type in here.

Not sure this is the cleanest approach. Alternatively I could extend the grouper and add yet another predicate type.

Reply:
I think a new attribute is cleaner and more flexible later if we also implement smarter grouping policy when # of pending tasks go up. 
For now those two approaches should be the same, so I am fine either way. 

(The task type predicate should already exist)

Reply:
Task type predicate is an enum, we can't use that because we're taking the HSM task type, which is an int32.
I'd have to make a new one.

I agree what I did isn't the cleanest approach.
I'll take another stab at it.